nncf/pruning/model_analysis.py

```diff
@@ -0,0 +1,243 @@
+
+def get_position(nx_nodes_list, idx):
+    for i, node in enumerate(nx_nodes_list):
+        if node['id'] == idx:
+            return i
+    return None
+
+
+def unit_clusters_for_nodes(nodes_to_merge, clusterization):
+    """
+
+    :param nodes_to_merge: all nodes are clusters for which should be тerged
+    :param clusterization:
+    """
+    if len(nodes_to_merge) <= 1:
+        return
+
+    # Will merge cluster with highest importance with others pairwise
+    max_importance_node_id = None
+    max_importance = 0
+    for node in nodes_to_merge:
+        importance = clusterization.get_cluster_for_node(node.node_id).importance
+        if importance > max_importance:
+            max_importance_node_id = node.node_id
+            max_importance = importance
+
+    max_importance_cluster_id = clusterization.get_cluster_for_node(max_importance_node_id).id
+    for node in nodes_to_merge:
+        if node.node_id != max_importance_node_id:
+            current_node_cluster_id = clusterization.get_cluster_for_node(node.node_id).id
+            clusterization.merge_clusters(max_importance_cluster_id, current_node_cluster_id)
+
+
+def cluster_special_ops_in_model(model: object, special_types: object, identity_types: object) -> Clusterization:
+    """
+    This model will cluster all operations with type from special_types. Connected nodes is nodes that:
+        1. Have path between nodes with only identity type nodes on it
+        2. Have common input (identity type nodes can be on path from this input)
+    :param model:
+    :param special_types:
+    :return: list of lists with clusters of special_types nodes (of type NNCFNode)
+    """
+    graph = model.get_original_graph()
+    nx_graph = graph._nx_graph
+    topologically_sorted_nodes = [nx_graph.nodes[node_name] for node_name in nx.topological_sort(nx_graph)]
+    all_special_nodes = [nx_graph.nodes[node_name] for node_name in nx_graph.nodes
+                         if graph.node_type_fn(nx_graph.nodes[node_name]) in special_types]
+
+    # 0. Initially all nodes is a separate clusters
+    clusterization = Clusterization("node_id")
+    for i, node in enumerate(all_special_nodes):
+        nncf_node = graph._nx_node_to_nncf_node(node)
+        cluster = NodesCluster(i, [nncf_node], [get_position(topologically_sorted_nodes, node['id'])])
+        clusterization.add_cluster(cluster, i)
+
+    for node in topologically_sorted_nodes:
+        if graph.node_type_fn(node) in identity_types:
+            continue
+
+        nncf_node = graph._nx_node_to_nncf_node(node)
+
+        all_outputs = find_next_nodes_not_types(model, nncf_node, identity_types)
+        all_output_special_nodes = [node for node in all_outputs if node.op_exec_context.operator_name in special_types]
+        if graph.node_type_fn(node) in special_types:
+            all_output_special_nodes.append(nncf_node)
+        unit_clusters_for_nodes(all_output_special_nodes, clusterization)
+
+    return clusterization
+
```

Comment:
Perhaps `unit_clusters_for_nodes` and `cluster_special_ops_in_model` could be made instance methods of a class like `PruningAnalyzer`?

Reply:
I don't think so. `ModelAnalyzer` has a specific purpose: model analysis in order to mark potentially prunable paths in the model. Whereas `unit_clusters_for_nodes` and `cluster_special_ops_in_model` works with grouping of special (usually Eltwise nodes).